pkg/kv/store_message.go

```diff
@@ -51,95 +52,91 @@ func (s *StoreMessage) DeleteMsg(ctx context.Context, path string) error {
 	return s.Delete(ctx, []byte(path))
 }
 
-func (s *StoreMessage) Scan(ctx context.Context, prefix string) (*PrimaryIterator, error) {
-	return NewPrimaryIterator(ctx, s.Store, prefix)
+func (s *StoreMessage) Scan(ctx context.Context, msgType protoreflect.MessageType, prefix string) (*PrimaryIterator, error) {
+	return NewPrimaryIterator(ctx, s.Store, msgType, prefix, "")
 }
 
 type MessageEntry struct {
 	Key   string
 	Value protoreflect.ProtoMessage
 }
 
-func NewMessageEntry(msg protoreflect.MessageType) *MessageEntry {
-	return &MessageEntry{
-		Key:   "",
-		Value: msg.New().Interface(),
-	}
-}
-
 type MessageIterator interface {
 	Next() bool
-	Entry(key *string, value *protoreflect.ProtoMessage) error
+	// Entry Receives key and value pointers from user and populates them with the iterator entry values
+	// The function will try to populate the func arguments which are not nil
+	Entry() *MessageEntry
 	Err() error
 	Close()
 }
 
 // PrimaryIterator MessageIterator implementation for primary key
 // The iterator iterates over the given prefix and returns the proto message and key
 type PrimaryIterator struct {
-	itr EntriesIterator
-	err error
+	itr     EntriesIterator
+	msgType protoreflect.MessageType
+	err     error
 }
 
-func NewPrimaryIterator(ctx context.Context, store Store, prefix string) (*PrimaryIterator, error) {
-	itr, err := ScanPrefix(ctx, store, []byte(prefix))
+func NewPrimaryIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, prefix, after string) (*PrimaryIterator, error) {
+	itr, err := ScanPrefix(ctx, store, []byte(prefix), []byte(after))
 	if err != nil {
 		return nil, fmt.Errorf("create prefix iterator: %w", err)
 	}
-	return &PrimaryIterator{itr: itr}, nil
+	return &PrimaryIterator{itr: itr, msgType: msgType}, nil
 }
 
-func (m *PrimaryIterator) Next() bool {
-	if m.err != nil {
+func (i *PrimaryIterator) Next() bool {
+	if i.err != nil {
 		return false
 	}
-	return m.itr.Next()
+	return i.itr.Next()
 }
 
-func (m *PrimaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
-	entry := m.itr.Entry()
+func (i *PrimaryIterator) Entry() *MessageEntry {
+	entry := i.itr.Entry()
 	if entry == nil {
-		return ErrNotFound
+		return nil
 	}
-	if value != nil {
-		err := proto.Unmarshal(entry.Value, *value)
-		if err != nil {
-			return fmt.Errorf("unmarshal proto data for key %s: %w", string(entry.Key), err)
-		}
-	}
-	if key != nil {
-		*key = string(entry.Key)
+	value := i.msgType.New().Interface()
+	err := proto.Unmarshal(entry.Value, value)
```

Comment:
Shouldn't we move this logic to `Next`? 
The reason I'm asking is that the user calls `Next()` and gets true. Now he calls `Entry` and gets a `nil` with a new error. The iterator usage we're accustomed to is that when `Next()` returns `true` you would **always** have a value to read. This is how we assemble the iterators across the board.
To have the same functionality here, you would have to move the parsing code to `Next` and store the value.